test/services/TxSender.test.ts

```diff
@@ -0,0 +1,157 @@
+
+    describe('when transaction is submitted successfully', () => {
+      before(() => {
+        txSender = new TxSender({ wallet: fakeWallet, chainId: 'eth' });
+      });
+
+      //* This test is not exiting. This suggests
+      //* that one asynchronous operation is not being finalized.
+      it('resolves the transaction receipt', async () => {
+        const fakeTransactionFunction = makeFakeTransactionFunction({ status: 1 });
+        const subject = await txSender.apply(fakeTransactionFunction, 100, 500, .01);
+        expect(subject.status).to.be.equal(1);
+      });
+    });
+
+    describe('when transaction is not submitted successfully', () => {
+      beforeEach(() => {
+        // casting is necessary to ensure compatibility.
+        fakeWallet = new FakeWallet(fakeProvider) as unknown as Wallet;
+        txSender = new TxSender({ wallet: fakeWallet, chainId: 'eth' });
+      });
+
+      it('attempts to cancel the pending transaction', async () => {
+        const spy = sinon.spy(fakeWallet, 'getTransactionCount');
+        const fakeTransactionFunction = makeFakeTransactionFunction();
+
+        try {
+          await txSender.apply(fakeTransactionFunction, 100, 500, .01);
+        } catch {}
+
+        expect(spy.called).to.be.true;
+      });
+
+      it('throws a timeout error', async () => {
+        const fakeTransactionFunction = makeFakeTransactionFunction();
+
+        const subject = txSender.apply(fakeTransactionFunction, 100, 500, .01);
+        expect(subject).to.eventually.throw('mint TX timeout');
+      });
+
+      describe('when transaction is canceled successfully', () => {
+        it('does not try to re-send the cancelling transaction', async () => {
+          const expectedCallCount = 4;
+          const fakeTransactionFunction = makeFakeTransactionFunction();
+
+          try {
+            await txSender.apply(fakeTransactionFunction, 100, 500, .01);
+          } catch {}
+
+          expect(fakeProvider.getBlockNumber.callCount).to.eql(expectedCallCount);
+        });
+      });
+
+      describe('when transaction is not canceled successfully', () => {
+        describe('when replacement fee is too low', () => {
+          it('re-send the cancelling transaction with different gas', async () => {
+            const firstCancellationAttemptCall = 3;
+            const error = new Error('replacement fee too low');
+            fakeProvider.getBlockNumber.onCall(firstCancellationAttemptCall).rejects(error);
+
+            const expectedCallCount = 6;
+            const fakeTransactionFunction = makeFakeTransactionFunction();
+
+            try {
+              await txSender.apply(fakeTransactionFunction, 100, 500, .01);
+            } catch {}
+
+            expect(fakeProvider.getBlockNumber.callCount).to.eql(expectedCallCount);
+          });
+
+          describe('when last attempt fails', () => {
+            it('throws an error', () => {
+              const firstCancellationAttemptCall = 3;
```

Comment:
It is not clear why we need to mock the firstCancellationAttemptCall since the test it is for the last attempt.

Reply:
You can't get to the last attempt if the first one doesn't fail before.

It works like this:

- Tries to submit a transaction
- If it fails, try to submit a cancellation transaction
- If the cancellation fails, try to submit again.

Reply:
Got it